File: Heap/maxScore.py
from heapq import heappush, heappop  # برای استفاده از min-heap (صف اولویت با کمینه در بالای هرم)

class Solution(object):
    def maxScore(self, nums1, nums2, k):
        n = len(nums1)  # طول آرایه‌ها را می‌گیریم
        
        # گام 1: جفت‌ها را درست می‌کنیم و بر اساس nums2 نزولی مرتب می‌کنیم
        # هر جفت = (nums2[i], nums1[i])
        pairs = sorted(zip(nums2, nums1), reverse=True)
        # zip(nums2, nums1) -> [(nums2[0], nums1[0]), (nums2[1], nums1[1]), ...]
        # sorted(..., reverse=True) -> مرتب سازی نزولی بر اساس عضو اول هر جفت (nums2)
        
        min_heap = []  # اینجا بزرگترین‌ها از nums1 را نگه می‌داریم
        sum_nums1 = 0  # مجموع اعداد انتخاب شده از nums1
        max_score = 0  # بیشترین امتیاز پیدا شده تا الان

        # گام 2: پیمایش روی جفت‌ها
        for n2, n1 in pairs:  # هر جفت = (عضو از nums2، عضو از nums1)
            heappush(min_heap, n1)  # اضافه کردن n1 به heap
            sum_nums1 += n1          # جمع را به‌روزرسانی می‌کنیم

            # اگر بیش از k عضو داریم، کوچکترین را حذف می‌کنیم
            if len(min_heap) > k:
                removed = heappop(min_heap)  # کوچکترین عضو را از heap خارج می‌کنیم
                sum_nums1 -= removed          # جمع را به‌روزرسانی می‌کنیم

            # اگر دقیقاً k عضو داریم → امتیاز را حساب می‌کنیم
            if len(min_heap) == k:
                score = sum_nums1 * n2       # n2 چون مرتب نزولی است، کمینه nums2 فعلی است
                max_score = max(max_score, score)  # بیشترین امتیاز را ذخیره می‌کنیم

        return max_score

File: Heap/maxScore.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number}
 */

// my first appeoach that didn't work
var maxScore = function (nums1, nums2, k) {
  // let score = (nums1[i0] + nums1[i1] +...+ nums1[ik - 1]) * min(nums2[i0] , nums2[i1], ... ,nums2[ik - 1]).
  let j = 0;
  let result = [];
  while (j + k < nums1.length) {
    let sum = 0;
    let min = Infinity;
    for (let i = 0; i < k; i++) {
      sum += nums1[i];
      min = Math.min(min, nums2[i]);
    }
    result.push(sum * min);
    j++;
  }
  return Math.max(result);
};

// nums1 = [1,3,3,2], nums2 = [2,1,3,4], k = 3

var maxScore = function (nums1, nums2, k) {
  const n = nums1.length;
  // [nums2, nums1] درست می‌کنیم و بر اساس nums2 نزولی مرتب می‌کنیم
  const pairs = Array.from({ length: n }, (_, i) => [nums2[i], nums1[i]]).sort(
    (a, b) => b[0] - a[0]
  );

  let arr = []; // اینجا بزرگترین‌های nums1 رو نگه می‌داریم
  let sum = 0,
    ans = 0;

  for (const [n2, n1] of pairs) {
    arr.push(n1);
    sum += n1;

    // اگه بیشتر از k شد، کوچیکترین رو حذف کن
    if (arr.length > k) {
      // مرتب کن و کوچیکترین رو بنداز بیرون
      arr.sort((a, b) => a - b);
      const removed = arr.shift();
      sum -= removed;
    }

    // اگه دقیقاً k عنصر داریم → کاندید جواب
    if (arr.length === k) {
      ans = Math.max(ans, sum * n2);
    }
  }

  return ans;
};

var maxScore = function (nums1, nums2, k) {
  // we should create a pair of nums1 and nums2 elements and sort it based on the nums2
  const pairs = nums2
    .map((n2, i) => [n2, nums1[i]])
    .sort((a, b) => b[0] - a[0]);
  // const n = nums1.length
  // const pairs = Array.from({length:n}, (_,i)=> [nums2[i],nums1[i]]).sort((a,b)=> b[0]-a[0])
  let minHeap = [];
  let sum = 0;
  let res = 0;

  for (let [n2, n1] of pairs) {
    sum += n1;
    minHeap.push(n1);

    if (minHeap.length > k) {
      minHeap.sort((a, b) => a - b);
      let removed = minHeap.shift();
      sum -= removed;
    }
    if (minHeap.length === k) {
      res = Math.max(res, sum * n2);
    }
  }
  return res;
};

// nums1 = [4,2,3,1,1,13],
// nums2 = [7,5,10,9,6,8], k = 3

// pairs = [[10,3], [9,1], [8,13], [7,4], [6,1], [5,2]]
// minHeap(sorted) = [3,4,13]
// sum = 20
// [(8*17), (7*20)]
// [136, 140 ]
// result = 140
